%\documentclass[10pt,a4paper]{article}
%\documentclass[twocolumn]{\VAR{bmcard}}  % uncomment this for twocolumn layout and comment line below

%uncomment this for twocolumn layout and comment line below
\documentclass[10pt]{\VAR{bmcard}} % uncomment this for twocolumn layout and comment line below

\usepackage{tabularx}
\usepackage[english]{babel}
\usepackage{array}
\newcolumntype{M}{>{\centering\arraybackslash}m{\dimexpr.25\linewidth-2\tabcolsep}}

\usepackage{color}

\usepackage{caption}
\usepackage{fancyhdr}
\usepackage{booktabs}
\usepackage{hyperref}

\usepackage[top=2in, bottom=1.5in, left=1in, right=1in]{geometry}

%\renewcommand{\headheight}{0.6in}
\setlength{\headwidth}{\textwidth}
\fancyhead[L]{ % right
%$   \includegraphics[height=0.53in]{\VAR{logo}}
ChiLin \VAR{version}
}
\fancyhead[R]{
prepared for: \VAR{user}
}% empty left
\pagestyle{fancy}

%\usepackage[paperwidth=15in, paperheight=30in]{geometry} %% margin=1.5in
%\usepackage{fullpage}
%\usepackage[cm]{fullpage}
%\usepackage[top=2.5in, bottom=1.5in, left=1in, right=2.5in]{geometry}
\usepackage{graphicx}
\usepackage{float}
\restylefloat{table}
\restylefloat{figure}
\usepackage[utf8]{inputenc} %unicode support


%%% Begin ...
\begin{document}
\begin{figure}
\raggedright
\vspace{-1in}\includegraphics[height=0.53in]{\VAR{logo}}
\end{figure}
%$\begin{frontmatter}
%
%$\begin{fmbox}
%\dochead{Softwares}
\begin{center}
{\huge Chip-Seq QC Report For ``\VAR{prefix_dataset_id}''}
\\*
{\large \today}
\end{center}

%$\author[
%$email={prepared for: \VAR{user}}
%$]{\inits{}\fnm{{\huge ChiLin \VAR{version}}\\*{\small Center for Cancer Epigenetics\\*Dana-Farber Cancer Institute}}\snm{}}
%% \author[
%% addressref={aff1},                   % id's of addresses, e.g. {aff1,aff2}
%% %corref={Brooklines 450},                       % id of corresponding address, if any
%% % noteref={n1},                        % id's of article notes, if any
%% %email={xsliu.dfci@gmail.com}   % email address
%% ]{\inits{CFCE }\fnm{}\snm{}}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%                                          %%
%%% Enter the authors' addresses here        %%
%%%                                          %%
%%% Repeat \address commands as much as      %%
%%% required.                                %%
%%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% %%
%
%% \address[id=aff1]{%                        % unique id
%%   \orgname{College of Life science and technology, Tongji}, % university, etc
%%   \street{Siping Road},                    %
%%   %\postcode{}                             % post or zip code
%%   \city{Shanghai},                         % city
%%   \cny{China}                              % country
%% }
%
%% \begin{artnotes}
%% \note[id=n1]{Equal contributor} % note, connected to author
%% \end{artnotes}
%
%$\end{fmbox}% comment this for two column layout
%\begin{abstractbox}
%
%\begin{abstract} % abstract
%\parttitle{ChIP-seq QC} %if any
%Chromatin immunoprecipitation (ChIP) followed by high-throughput DNA sequencing (ChIP-seq) has become a valuable and widely used approach for mapping the genomic location of transcription-factor binding and histone modifications in living cells. Thousands Chip-seq data are generated by different lab, ChinLin quality control aims to  score and evaluate Chip-seq data experiment and analysis quality based on our collected dataset, include raw data QC, reads mapping QC, peak calling QC and a series of annotation QC.
%% Text for this section.
%% \parttitle{Second part title} %if any
%% Text for this section.
%\end{abstract}
%% \begin{keyword}
%% \kwd{sample}
%% \kwd{article}
%% \kwd{author}
%% \end{keyword}
%\end{abstractbox}
%$\end{frontmatter}
